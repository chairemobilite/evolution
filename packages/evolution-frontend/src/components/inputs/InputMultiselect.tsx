/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { WithTranslation, withTranslation } from 'react-i18next';
import Select from 'react-select';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import * as surveyHelper from 'evolution-common/lib/utils/helpers';
import { ChoiceType, InputMultiselectType } from 'evolution-common/lib/services/widgets';
import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import { CliUser } from 'chaire-lib-common/lib/services/user/userType';

interface InputMultiselectProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> {
    id: string;
    onValueChange?: (e: any) => void;
    value: string[];
    interview: UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    // TODO There's also a path in widgetConfig, but this one comes from the props of the question. See if it's always the same and use the one from widgetConfig if necessary
    path: string;
    user: CliUser;
    inputRef?: React.LegacyRef<HTMLInputElement>;
    widgetConfig: InputMultiselectType<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
}

const formatOptionLabel = (data: { color?: string; label: string }) => (
    <div style={{ display: 'flex', alignItems: 'center' }}>
        <div
            style={{
                background: data.color,
                borderRadius: 6,
                height: 6,
                width: 6,
                marginRight: 6
            }}
        />
        {data.label}
    </div>
);

type OptionType = { label: string; color?: string; value: string };

export class InputMultiselect<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> extends React.Component<
    InputMultiselectProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> & WithTranslation
> {
    constructor(props) {
        super(props);
        this.onValueChange = this.onValueChange.bind(this);
        this.selectShortcut = this.selectShortcut.bind(this);
        this.customFilter = this.customFilter.bind(this);
    }

    onValueChange = (valuesArray) => {
        if (this.props.onValueChange === undefined) {
            return;
        }
        if (valuesArray) {
            if (this.props.widgetConfig.multiple !== false) {
                if (!Array.isArray(valuesArray)) {
                    valuesArray = [valuesArray];
                }
                this.props.onValueChange({ target: { value: valuesArray.map((option) => option.value) } });
            } else {
                this.props.onValueChange({ target: { value: valuesArray.value } });
            }
        } else {
            this.props.onValueChange({ target: { value: null } });
        }
    };

    selectShortcut = (
        value: ChoiceType<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>,
        e: React.MouseEvent
    ) => {
        if (this.props.onValueChange === undefined) {
            return;
        }
        if (e) {
            e.preventDefault();
        }
        if (this.props.widgetConfig.multiple !== false) {
            const valuesArray = _isBlank(this.props.value) ? [] : this.props.value.slice(0); // shallow array clone
            valuesArray.push(value.value);
            this.props.onValueChange({ target: { value: valuesArray } });
        } else {
            this.props.onValueChange({ target: { value: value.value } });
        }
    };

    customFilter = (option: OptionType, searchText: string) => {
        if (option.label.toLowerCase().includes(searchText.toLowerCase())) {
            return true;
        } else {
            return false;
        }
    };

    render() {
        const shortcutButtons: JSX.Element[] = [];
        let hasShortcuts = false;
        const actualValue = _isBlank(this.props.value) ? [] : this.props.value;
        const actualValueToSort: { choice: OptionType; index: number }[] = [];
        const choices =
            typeof this.props.widgetConfig.choices === 'function'
                ? this.props.widgetConfig.choices(this.props.interview, this.props.path)
                : this.props.widgetConfig.choices;
        const selectOptions = choices
            .filter((choice) => {
                return (
                    choice.hidden !== true &&
                    surveyHelper.parseBoolean(
                        choice.conditional,
                        this.props.interview,
                        this.props.path,
                        this.props.user
                    )
                );
            })
            .map((choice) => {
                const label = surveyHelper.translateString(
                    choice.label,
                    this.props.i18n,
                    this.props.interview,
                    this.props.path,
                    this.props.user
                );
                const current = {
                    label: label as string,
                    color: choice.color,
                    value: choice.value
                };
                if (actualValue.indexOf(choice.value) > -1) {
                    actualValueToSort.push({
                        choice: current,
                        index: actualValue.indexOf(choice.value)
                    });
                }
                return current;
            }, this);

        const actualValueWithLabel = actualValueToSort
            .sort((valueObjectA, valueObjectB) => {
                return valueObjectA.index - valueObjectB.index;
            })
            .map((v2s) => v2s.choice);

        if (this.props.widgetConfig.shortcuts) {
            hasShortcuts = true;
            for (let i = 0, count = this.props.widgetConfig.shortcuts.length; i < count; i++) {
                const shortcut = this.props.widgetConfig.shortcuts[i];
                shortcutButtons.push(
                    <button
                        key={'shortcut' + i}
                        type="button"
                        className={`button shortcut-button${shortcut.color ? ` ${shortcut.color}` : 'blue'}`}
                        onClick={(e) => this.selectShortcut(shortcut, e)}
                        tabIndex={-1}
                    >
                        {shortcut.icon && <FontAwesomeIcon icon={shortcut.icon} className="faIconLeft" />}
                        {surveyHelper.translateString(
                            shortcut.label,
                            this.props.i18n,
                            this.props.interview,
                            this.props.path,
                            this.props.user
                        )}
                    </button>
                );
            }
        }

        const customStyles = {
            option: (base, state) => ({
                ...base,
                padding: '0.25rem 0.5rem'
            }),
            menu: (base, state) => ({
                ...base,
                marginTop: '0'
            })
        };

        // TODO: maybe we should jsut extend react select and accept any of its config in our component props
        return (
            <div className="survey-question__input-multiselect-container">
                <Select
                    formatOptionLabel={formatOptionLabel}
                    //disabled={false}
                    isMulti={this.props.widgetConfig.multiple !== false}
                    onChange={this.onValueChange}
                    options={selectOptions}
                    filterOption={this.props.widgetConfig.onlyLabelSearch ? this.customFilter : undefined}
                    isSearchable={true}
                    isClearable={this.props.widgetConfig.isClearable || false}
                    closeMenuOnSelect={
                        this.props.widgetConfig.closeMenuOnSelect || (this.props.widgetConfig.multiple ? false : true)
                    }
                    placeholder=""
                    name={`survey-question__input-multiselect-${this.props.path}`}
                    value={this.props.widgetConfig.multiple !== false ? actualValueWithLabel : actualValueWithLabel[0]}
                    className="react-select-container"
                    classNamePrefix="react-select"
                    styles={customStyles}
                />
                {hasShortcuts && shortcutButtons}
            </div>
        );
    }
}

export default withTranslation()(InputMultiselect) as React.FunctionComponent<
    InputMultiselectProps<any, any, any, any>
>;
