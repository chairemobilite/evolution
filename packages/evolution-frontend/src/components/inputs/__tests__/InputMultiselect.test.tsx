/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import TestRenderer from 'react-test-renderer';

import { interviewAttributes } from './interviewData.test';
import InputMultiselect from '../InputMultiselect';

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
}

describe('Render InputMultiselect with various options', () => {

    const conditionalFct = jest.fn().mockReturnValue(true);
    const choices = [
        {
            value: 'val1',
            label: { en: 'english value', fr: 'valeur française' },
            hidden: false,
            icon: 'creative-commons' as const
        },
        {
            value: 'val2',
            label: 'Unilingual label',
            iconPath: 'img/test.png'
        },
        {
            value: 'hiddenVal',
            label: { en: 'english hidden', fr: 'cachée' },
            hidden: true
        },
        {
            value: 'conditionalVal',
            label: { en: 'english conditional', fr: 'conditionnelle' },
            conditional: conditionalFct
        }
    ];

    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        inputType: 'multiselect' as const,
        choices,
        containsHtml: true,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    };

    test('Basic options', () => {
        const wrapper = TestRenderer.create(
            <InputMultiselect
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={widgetConfig}
                value={['val1','val2']}
                inputRef={React.createRef()}
                size='medium'
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
        expect(conditionalFct).toHaveBeenCalledTimes(1);
        expect(conditionalFct).toHaveBeenCalledWith(interviewAttributes, 'foo.test', userAttributes);
    });
    
    test('With shortcuts', () => {
        const configWithShortcuts = Object.assign({}, widgetConfig, { 
            shortcuts: [
                {
                    value: 'val2',
                    label: 'Unilingual label',
                    iconPath: 'img/test.png'
                },
                {
                    value: 'conditionalVal',
                    label: { en: 'english conditional', fr: 'conditionnelle' },
                    color: 'red'
                },
            ]
        });
        const wrapper = TestRenderer.create(
            <InputMultiselect
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={configWithShortcuts}
                value={['val1']}
                inputRef={React.createRef()}
                size='medium'
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('With all parameters', () => {
        const configWithShortcuts = Object.assign({}, widgetConfig, { 
            shortcuts: [
                {
                    value: 'val2',
                    label: 'Unilingual label',
                    iconPath: 'img/test.png'
                },
                {
                    value: 'conditionalVal',
                    label: { en: 'english conditional', fr: 'conditionnelle' },
                    color: 'red'
                },
            ],
            multiple: true, // Should have only one value
            onlyLabelSearch: true,
            isClearable: true,
            closeMenuOnSelect: true
        });
        const wrapper = TestRenderer.create(
            <InputMultiselect
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={configWithShortcuts}
                value={['val1', 'val2']}
                inputRef={React.createRef()}
                size='medium'
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
    
});
