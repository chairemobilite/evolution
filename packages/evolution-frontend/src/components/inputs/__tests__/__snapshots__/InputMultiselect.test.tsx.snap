// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render InputMultiselect with various options Basic options 1`] = `
<div
  className="survey-question__input-multiselect-container"
>
  <div
    className="react-select-container css-b62m3t-container"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-2-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className="react-select__control css-13cymwt-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="react-select__value-container react-select__value-container--is-multi react-select__value-container--has-value css-3w2yfm-ValueContainer"
      >
        <div
          className="react-select__multi-value css-1p3m7a8-multiValue"
        >
          <div
            className="react-select__multi-value__label css-wsp0cs-MultiValueGeneric"
          >
            <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                }
              }
            >
              <div
                style={
                  Object {
                    "background": undefined,
                    "borderRadius": 6,
                    "height": 6,
                    "marginRight": 6,
                    "width": 6,
                  }
                }
              />
              english value
            </div>
          </div>
          <div
            aria-label="Remove [object Object]"
            className="react-select__multi-value__remove css-12a83d4-MultiValueRemove"
            onClick={[Function]}
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            role="button"
          >
            <svg
              aria-hidden="true"
              className="css-tj5bde-Svg"
              focusable="false"
              height={14}
              viewBox="0 0 20 20"
              width={14}
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </div>
        </div>
        <div
          className="react-select__multi-value css-1p3m7a8-multiValue"
        >
          <div
            className="react-select__multi-value__label css-wsp0cs-MultiValueGeneric"
          >
            <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                }
              }
            >
              <div
                style={
                  Object {
                    "background": undefined,
                    "borderRadius": 6,
                    "height": 6,
                    "marginRight": 6,
                    "width": 6,
                  }
                }
              />
              Unilingual label
            </div>
          </div>
          <div
            aria-label="Remove [object Object]"
            className="react-select__multi-value__remove css-12a83d4-MultiValueRemove"
            onClick={[Function]}
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            role="button"
          >
            <svg
              aria-hidden="true"
              className="css-tj5bde-Svg"
              focusable="false"
              height={14}
              viewBox="0 0 20 20"
              width={14}
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </div>
        </div>
        <div
          className="react-select__input-container css-qbdosj-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={true}
            aria-labelledby="test_label"
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className="react-select__input"
            disabled={false}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="react-select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <span
          className="react-select__indicator-separator css-1u9des2-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className="react-select__indicator react-select__dropdown-indicator css-1xc3v61-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-tj5bde-Svg"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div>
      <input
        name="survey-question__input-multiselect-foo.test"
        type="hidden"
        value="val1"
      />
      <input
        name="survey-question__input-multiselect-foo.test"
        type="hidden"
        value="val2"
      />
    </div>
  </div>
</div>
`;

exports[`Render InputMultiselect with various options With all parameters 1`] = `
<div
  className="survey-question__input-multiselect-container"
>
  <div
    className="react-select-container css-b62m3t-container"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-4-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className="react-select__control css-13cymwt-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="react-select__value-container react-select__value-container--is-multi react-select__value-container--has-value css-3w2yfm-ValueContainer"
      >
        <div
          className="react-select__multi-value css-1p3m7a8-multiValue"
        >
          <div
            className="react-select__multi-value__label css-wsp0cs-MultiValueGeneric"
          >
            <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                }
              }
            >
              <div
                style={
                  Object {
                    "background": undefined,
                    "borderRadius": 6,
                    "height": 6,
                    "marginRight": 6,
                    "width": 6,
                  }
                }
              />
              english value
            </div>
          </div>
          <div
            aria-label="Remove [object Object]"
            className="react-select__multi-value__remove css-12a83d4-MultiValueRemove"
            onClick={[Function]}
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            role="button"
          >
            <svg
              aria-hidden="true"
              className="css-tj5bde-Svg"
              focusable="false"
              height={14}
              viewBox="0 0 20 20"
              width={14}
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </div>
        </div>
        <div
          className="react-select__multi-value css-1p3m7a8-multiValue"
        >
          <div
            className="react-select__multi-value__label css-wsp0cs-MultiValueGeneric"
          >
            <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                }
              }
            >
              <div
                style={
                  Object {
                    "background": undefined,
                    "borderRadius": 6,
                    "height": 6,
                    "marginRight": 6,
                    "width": 6,
                  }
                }
              />
              Unilingual label
            </div>
          </div>
          <div
            aria-label="Remove [object Object]"
            className="react-select__multi-value__remove css-12a83d4-MultiValueRemove"
            onClick={[Function]}
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            role="button"
          >
            <svg
              aria-hidden="true"
              className="css-tj5bde-Svg"
              focusable="false"
              height={14}
              viewBox="0 0 20 20"
              width={14}
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </div>
        </div>
        <div
          className="react-select__input-container css-qbdosj-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={true}
            aria-labelledby="test_label"
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className="react-select__input"
            disabled={false}
            id="react-select-4-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="react-select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          className="react-select__indicator react-select__clear-indicator css-1xc3v61-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-tj5bde-Svg"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
        <span
          className="react-select__indicator-separator css-1u9des2-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className="react-select__indicator react-select__dropdown-indicator css-1xc3v61-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-tj5bde-Svg"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div>
      <input
        name="survey-question__input-multiselect-foo.test"
        type="hidden"
        value="val1"
      />
      <input
        name="survey-question__input-multiselect-foo.test"
        type="hidden"
        value="val2"
      />
    </div>
  </div>
  <button
    className="button shortcut-buttonblue"
    onClick={[Function]}
    tabIndex={-1}
    type="button"
  >
    Unilingual label
  </button>
</div>
`;

exports[`Render InputMultiselect with various options With shortcuts 1`] = `
<div
  className="survey-question__input-multiselect-container"
>
  <div
    className="react-select-container css-b62m3t-container"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-3-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className="react-select__control css-13cymwt-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="react-select__value-container react-select__value-container--is-multi react-select__value-container--has-value css-3w2yfm-ValueContainer"
      >
        <div
          className="react-select__multi-value css-1p3m7a8-multiValue"
        >
          <div
            className="react-select__multi-value__label css-wsp0cs-MultiValueGeneric"
          >
            <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                }
              }
            >
              <div
                style={
                  Object {
                    "background": undefined,
                    "borderRadius": 6,
                    "height": 6,
                    "marginRight": 6,
                    "width": 6,
                  }
                }
              />
              english value
            </div>
          </div>
          <div
            aria-label="Remove [object Object]"
            className="react-select__multi-value__remove css-12a83d4-MultiValueRemove"
            onClick={[Function]}
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            role="button"
          >
            <svg
              aria-hidden="true"
              className="css-tj5bde-Svg"
              focusable="false"
              height={14}
              viewBox="0 0 20 20"
              width={14}
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </div>
        </div>
        <div
          className="react-select__input-container css-qbdosj-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={true}
            aria-labelledby="test_label"
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className="react-select__input"
            disabled={false}
            id="react-select-3-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="react-select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <span
          className="react-select__indicator-separator css-1u9des2-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className="react-select__indicator react-select__dropdown-indicator css-1xc3v61-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-tj5bde-Svg"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div>
      <input
        name="survey-question__input-multiselect-foo.test"
        type="hidden"
        value="val1"
      />
    </div>
  </div>
  <button
    className="button shortcut-buttonblue"
    onClick={[Function]}
    tabIndex={-1}
    type="button"
  >
    Unilingual label
  </button>
  <button
    className="button shortcut-button red"
    onClick={[Function]}
    tabIndex={-1}
    type="button"
  >
    english conditional
  </button>
  <button
    className="button shortcut-button red"
    onClick={[Function]}
    tabIndex={-1}
    type="button"
  >
    Translated shortcut
  </button>
</div>
`;
